requests/sessions.py

```diff
@@ -242,7 +242,9 @@ def rebuild_auth(self, prepared_request, response):
             original_parsed = urlparse(response.request.url)
             redirect_parsed = urlparse(url)
 
-            if (original_parsed.hostname != redirect_parsed.hostname):
+            if (original_parsed.hostname != redirect_parsed.hostname
+                    or original_parsed.port != redirect_parsed.port
+                    or original_parsed.scheme != redirect_parsed.scheme):
```

Comment:
I think this might need to be a bit different. I think if someone sent credentials over `HTTP` and we're being redirected to the same domain but it's `HTTPS` we should allow that. That's a secure upgrade. The opposite should not be allowed.

Reply:
I did it this way based on RFC 7235:

>  A protection space is defined by the canonical root URI (the scheme
   and authority components of the effective request URI; see Section
   5.5 of [RFC7230]) of the server being accessed, in combination with
   the realm value if present.

...

> The protection space determines the domain over which credentials can
   be automatically applied.  If a prior request has been authorized,
   the user agent MAY reuse the same credentials for all other requests
   within that protection space for a period of time determined by the
   authentication scheme, parameters, and/or user preferences (such as a
   configurable inactivity timeout).

While it's not explicit, the implication is that a user agent may not automatically reuse the credentials outside the protection space, and a change from http -> https is a change in protection space.

Reply:
@bmerry, just to be clear, the spec you’re referencing is specifically about the realm attribute, which is different that what we’re doing here.

If we look at browsers for reference here, upgrade redirects maintain Authorization headers. Breaking this will make a common pattern pretty painful for most users.

Reply:
Strange, I thought I'd tested the http -> https case with Firefox and Chromium and neither had sent the Authorization header. I'll take a closer look at my test when I'm in the office tomorrow.

Just to check that I'm testing the same thing that you're referring to: my test is a localhost Apache server listening on http and https. On http it is set up to require basic auth, and on some path it is set up to send a 302 redirect to somewhere on https. I point the browser at the http port, I get a WWW-Authenticate challenge and I type in username+password, I then get redirected to https and the browser either does or doesn't automatically send the same credentials immediately without another WWW-Authenticate challenge?

Reply:
I've redone my test, using `openssl s_server` to see the request headers that the browser sends to the https port after the redirect. Firefox definitely isn't preserving the Authorization header across the redirect. Let me know if you'd like more details to reproduce the test (httpd.conf etc) using Docker.

Reply:
Hi @bmerry, it looks like Firefox does adhere to this rule. I'm a bit worried that adding the stripping from HTTP->HTTPS is going to be breaking for some people. I think for this patch, we should focus on removing the immediate security risk, and then we can add a todo for Requests3 to make this more in line with how browsers are behaving.

Reply:
Thanks for reminding me about this PR. Okay, I can add an exception to allow http -> https redirects, with a comment to explain why.

How do you want to handle port changes? For example, `http://example.com` -> `https://example.com` is one thing (which implicitly changes the port from 80 to 443), but what about `http://example.com:8080` -> `https://example.com:12345` or `http://example.com` -> `http://example.com:12345`? That last one seems like a security risk since a non-root user could listen on an arbitrary port.

I bring up the port changes because the unit test necessarily runs httpbin on non-standard ports and so there is no way to test redirects with implicit port numbers.

Reply:
Any change in a URLs origin should probably result in stripping of security headers. Origin for `http/https` URLs is scheme, hostname, and port. The only cases we should keep headers on a cross-origin redirect IMO are `http://example.com:80 -> https://example.com:443` and `http://example.com:1234 -> https://example.com:1234`. Thoughts?

Reply:
I think that matches the logic I started implementing (with the exception that it will only allow upgrade from 80->443 when both are implicit):
```python
https_upgrade = original_parsed.scheme == 'http' and redirect_parsed.scheme == 'https'
if (original_parsed.hostname != redirect_parsed.hostname
         or original_parsed.port != redirect_parsed.port
         or (original_parsed.scheme != redirect_parsed.scheme and not https_upgrade)):
```
Does that look reasonable?

The catch is that it's not really possible to test with httpbin, because we don't get to pick the port numbers that httpbin runs on. One option would be to move this logic into its own function (that just takes two URLs and decides whether to strip) and unit test that function separately from the end-to-end test through httpbin. The existing end-to-end tests should be sufficient to check that the actual stripping is done correctly.

Reply:
Question because I don't know the behavior of the URL parser, if you parse `http://example.com` and `http://example.com:80` do you get a different value of `redirect_parsed.port`, is it `None` on implicit?

Reply:
Yes
```python
>>> print(urllib.parse.urlparse('http://example.com/').port)
None
>>> print(urllib.parse.urlparse('http://example.com:80/').port)
80
```

Reply:
So we'd have to account for that in our comparison above as we want `http://example.com:80` -> `https://example.com` to work the same as `http://example.com` -> `https://example.com`.

Reply:
At this point there’s enough complexity in this implementation we should probably move it into it’s own check function.

I think Seth’s assessment is right here. We want to check the following:

1. The hostnames match.
2. The schemes match...
    * Except in the case of http->https
    * If scheme change is http->https, origin port may be 80 or None and destination port may be 443 or None.
3. Ports match, except in above scenario.
4. Same scheme and hostname but varied port results in stripped auth.

Reply:
Sounds sensible to me. I've implemented that in the latest push.

I've left the logic in a public method of SessionRedirectMixin, because that gives the user the option to override it in a subclass, but let me know if I should move it to utils.

Reply:
Ping? I think I've addressed the comments so far.